mm/map/mm_map.c

```diff
@@ -0,0 +1,337 @@
+ *
+ ****************************************************************************/
+
+FAR struct mm_map_entry_s *mm_map_find_contains(FAR const void *vaddr)
+{
+  FAR struct tcb_s *tcb = nxsched_self();
+  FAR struct task_group_s *group = tcb->group;
+  FAR struct mm_map_s *mm = &group->tg_mm_map;
+  FAR struct mm_map_entry_s *map = NULL;
+
+  if (nxmutex_lock(&mm->mm_map_mutex) == OK)
+    {
+      map = (struct mm_map_entry_s *)sq_peek(&mm->mm_map_sq);
+
+      while (map && !vaddr_in_area(vaddr, map->vaddr, map->length))
+        {
+          map = (struct mm_map_entry_s *)sq_next(((sq_entry_t *)map));
+        }
+
+      nxmutex_unlock(&mm->mm_map_mutex);
+    }
+
+  return map;
+}
+
+/****************************************************************************
+ * Name: mm_map_find
+ *
+ * Description:
+ *   Find the first mapping matching address and length
+ *
+ ****************************************************************************/
+
+FAR struct mm_map_entry_s *mm_map_find(FAR const void *vaddr,
+                                             size_t length)
+{
+  FAR struct tcb_s *tcb = nxsched_self();
+  FAR struct task_group_s *group = tcb->group;
+  FAR struct mm_map_s *mm = &group->tg_mm_map;
+  FAR struct mm_map_entry_s *map = NULL;
+
+  if (nxmutex_lock(&mm->mm_map_mutex) == OK)
+    {
+      map = (struct mm_map_entry_s *)sq_peek(&mm->mm_map_sq);
+
+      while (map && vaddr != map->vaddr && length != map->length)
+        {
+          map = (struct mm_map_entry_s *)sq_next(((sq_entry_t *)map));
+        }
+
+      nxmutex_unlock(&mm->mm_map_mutex);
+    }
+
+  return map;
+}
+
+/****************************************************************************
+ * Name: mm_map_rm
+ *
+ * Description:
+ *   Remove a mapping from the task  group's list
+ *
+ ****************************************************************************/
+
+int mm_map_rm(FAR struct task_group_s *group,
+              FAR struct mm_map_entry_s **map)
+{
+  FAR struct mm_map_entry_s *prev;
+  FAR struct mm_map_entry_s *r = NULL;
+  FAR struct mm_map_s *mm;
+
+  if (!group)
+    {
+      /* We end up in here only through mm_map_destroy. The address
+       * environment is being destroyed and the group is not valid.
+       * just delete the entry
+       */
+
+      goto free_mem;
+    }
+
+  mm = &group->tg_mm_map;
+
+  int ret = nxmutex_lock(&mm->mm_map_mutex);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  prev = (struct mm_map_entry_s *)sq_peek(&mm->mm_map_sq);
+
+  /* Check if the list was empty */
+
+  if (!prev)
+    {
+      nxmutex_unlock(&mm->mm_map_mutex);
+      return -ENOENT;
+    }
+
+  /* Check if removing the first item */
+
+  if (*map == prev)
+    {
+      sq_remfirst(&mm->mm_map_sq);
+      *map = NULL;
+      r = prev;
+    }
+  else
+    {
+      /* Loop through the remaining items to find the one to be removed */
+
+      while ((r = (struct mm_map_entry_s *)sq_next(((sq_entry_t *)prev))))
```

Comment:
why not use dq_entry to avoid the loop

Reply:
just to save memory one pointer in every map_entry. I don't think that munmap is time critical (it is never real time anyhow).

Since munmap does need to always go through all the mappings it is algorithmically O(n^2) anyways, even if you used RBT here ( would become O(n) * O(log n) ) since even in that case you'd have to start search again after every found mapping in munmap.

Normally you'll anyways have just a couple of mathing mappings, so "typically" this is just linear; that is, in case of N matching entries you go through the list N+1 times. In normal case you map an area once, so with sq you got through the list twice, whereas with dq you'd go through it once. Not significant at least for me.

If someone, someday, wants to do e.g. RBT for this, it would consume 3 pointers + color.... At this point I don't see that necessary and it is somewhat complicated due to 2 keys (start and length)